---
output: github_document
---

<!-- README.md is generated from README.Rmd. Please edit that file -->

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "man/figures/README-",
  out.width = "100%"
)
```

# TRexDAD

A pipeline for the design of mutagenizing oligonucleotides required for Tile Region Exchange Mutagenesis (T-Rex), which is a crucial step for deep mutation scanning (DMS), by harnessing data-optimized assembly design (DAD).

<!-- badges: start -->

![GitHub](https://img.shields.io/github/license/yunyicheng/TRexDAD)
![GitHub language count](https://img.shields.io/github/languages/count/yunyicheng/TRexDAD)
![GitHub last commit (branch)](https://img.shields.io/github/last-commit/yunyicheng/TRexDAD/main)
![GitHub commit activity (branch)](https://img.shields.io/github/commit-activity/w/yunyicheng/TRexDAD/main)
![GitHub issues](https://img.shields.io/github/issues/yunyicheng/TRexDAD)
![Discord](https://img.shields.io/discord/1172273306162958441)

<!-- badges: end -->

## Description

Designing oligonucleotides is a significant part of Tile Region Exchange mutagenesis (T-Rex). However, derive the assembly design by hand is often error-prone and time-consuming since the fidelity of resulting oligonucleotides does not merely depend on mere rules. Therefore, a data-driven solution is required for designing oligonucleotides with high correct assembly rate. Although there are web tools currently available for this procedure, few of them utilized Data-optimized Assembly Design.`TRexDAD` is a streamline tool that facilitates oligonucleotide design for Tile Region Exchange Mutagenesis by harnessing Data-optimized Assembly Design (DAD) techniques, along with visualization of the results. The package is targeted for researchers who perform Tile Region Exchange Mutagenesis for some genes of interest and need an efficient and reliable streamline to produce mutagenizing oligonucleotides. The scope of the R package is to improve the work flow in Tile Region Exchange Mutagenesis, facilitating open-source researches for interested bioinformatitians and computational biologists. The `TRexDAD` package was developed using `R version 4.3.1 (2023-06-16)`, `Platform: aarch64-apple-darwin20 (64-bit)` and `Running under: macOS Sonoma 14.1.1`.

## Installation

To install the latest version of the package:

``` {r}
install.packages("devtools")
library("devtools")
devtools::install_github("yunyicheng/TRexDAD", build_vignettes = TRUE)
library("TRexDAD")
```

<!-- To run the Shiny app: -->

<!-- ``` r -->
<!-- runTestingPackage() # not for Assessment 4; only for Assessment 5 -->
<!-- ``` -->

## Package Overview

To list all user-accessible functions, run:
```{r}
ls("package:TRexDAD")
```

`TRexDAD` contains 10 functions.

1.  ***execute_and_plot*** is the work flow function which wraps up most of the other functions. It performs an optimization process to determine the optimal tile positions in a gene sequence and plots the progression of the score over iterations. It provides two customization parameters: `iteration_max` and `scan_rate`. If not other specified, these parameter will have default values: `iteration_max=30`, `scan_rate=7`.

2.  ***split_into_codons*** takes a gene sequence (as a string) and splits it into codons.

3.  ***oligo_cost*** calculates the cost of oligonucleotides (oligos) based on 
the number of tiles and the number of codons.

4. ***calculate_optimal_tiles*** calculates the oligo cost for a range of tile numbers and finds the number of tiles that minimizes this cost. It also computes the global length of the tiles.

5. ***get_overhangs*** calculates the overhang sequences at the specified positions in a gene sequence. Depending on the flag, it returns these sequences as either `DNAString` objects or plain character strings.

6. ***get_all_overhangs*** computes overhang sequences for a given list of positions in a gene sequence. It iteratively calls the `get_overhangs` function to calculate the head and tail overhangs for each position and accumulates them in a list.

7. ***obtain_score*** calculates a score for a specific tile within a gene sequence. The score is based on palindromicity, length variation from a global standard, and on-target reactivity, using a pre-defined overhang fidelity dataframe.

8. ***calculate_scores*** calculates a global score for a list of positions in a gene sequence. The score takes into account off-target reactions, repetitions, and the sum of local scores (by calling `obtain_score`). 
9. ***pick_position*** selects a position from a list of positions for optimization based on their scores. It uses a weighted sampling approach where the weights are inversely proportional to the scores of the positions.

10. ***optimize_position*** optimizes a single position within a list of positions. It adjusts the specified position to maximize the overall score (obtained via `calculate_scores`). The function supports different optimization modes, including greedy and Markov Chain Monte Carlo (MCMC) approaches.

## Workflow Usage
The main objective of the package is to find optimal assembly design in mutagenizing Rad27. The RAD27 gene of Saccharomyces cerevisiae encodes a 5′-3′ flap exo/endonuclease, which is significant for DNA replication.

To initiate workflow (with visualization of scores), run:
```{r}
execute_and_plot()
```

## Contributions

The author of the package is Yunyi Cheng. 
The author wrote the *execute_and_plot* function, which performs an optimization process to determine the optimal tile positions in a gene sequence and plots the progression of the score over iterations. 
Other user-accessible functions are also developed by the author. Most of them serve as helper functions for the *execute_and_plot* function
`Biostrings` R package is used for calculating reverse complement of overhangs in the *get_overhangs* function.
`readxl` R package is used for reading excel file `overhang_fidelity.xlsx` containing tested overhangs's reactivity against each other. `overhang_fidelity.xlsx` is from the study by New England Biolabs.
`forstringr` R package is used for manipulating splitted target gene (a list of codons) in the *get_overhangs* function.

The `stats` R package is used for optimizing gene tile length and number of tiles. It is used in  *calculate_optimal_tiles* function and *execute_and_plot* function. 


## References

- Vladimir Potapov, Jennifer L. Ong, Rebecca B. Kucera, Bradley W. Langhorst, Katharina Bilotti, John M. Pryor, Eric J. Cantor, Barry Canton, Thomas F. Knight, Thomas C. Jr. Evans, and Gregory J. S. Lohman. Comprehensive Profiling of Four Base Overhang Ligation Fidelity by T4 DNA Ligase and Application to DNA Assembly. ACS Synthetic Biology, 7(11):2665–2674, 2018.

- 

## Acknowledgements

This package was developed as part of an assessment for 2023 BCB410H: Applied Bioinformatics course at the University of Toronto, Toronto, CANADA. 

`TRexDAD` welcomes issues, enhancement requests, and other contributions. To submit an issue, use the [GitHub issues](https://github.com/yunyicheng/TRexDAD/issues). Many thanks to those who provided feedback to improve this package.


## Package Structure
```{r}
install(DependenciesGraph)
library(DependenciesGraph)
library(TRexDAD)
deps <- funDependencies("package:TRexDAD", "generate_split_coded_comments")
plot(deps)
```